### PR TITLE
[ONNX] Enable onnx shape inference in export by default

### PR DESCRIPTION
**Pull request by BowenBao**
* Enable ONNX shape inference by default.
* ONNX could potentially set inferred shape in output instead of value_infos, checking both to be sure.
* Small fix in symbol_map to avoid overlooking dup symbols.
* Fix scalar_type_analysis to be consistent with PyTorch scalar type promotion logic.
* Correctly handle None dim_param from ONNX inferred shape.

### CODE REVIEW

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 1]
```diff
@@ -3269,6 +3269,7 @@ def forward(self, x):
         self.run_test(Zero_(), x)
 
     @skipIfUnsupportedMinOpsetVersion(9)
+    @disableScriptTest()
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:3272]: note: this is failing the new scripting test, and is being fixed at another PR #45793

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 2]
```diff
@@ -3439,6 +3440,7 @@ def forward(self, x):
         self.run_test(MaskedSelectModel(), x)
 
     @skipIfUnsupportedMinOpsetVersion(11)
+    @disableScriptTest()
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:3443]: note: these two are failing the new scripting test, and is being fixed at #46866

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 3]
```diff
@@ -3452,6 +3454,7 @@ def forward(self, input_mask, some_const):
         self.run_test(MaskedFillModel(), (mask, constant))
 
     @skipIfUnsupportedMinOpsetVersion(11)
+    @disableScriptTest()
```
Comment (neginraoof) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:3457]: How about this test? Is this going to e addressed?

Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:3457]: yes this is also being addressed in #46866

File: torch/csrc/jit/passes/onnx/scalar_type_analysis.cpp [hunk 4]
```diff
@@ -108,8 +108,29 @@ static c10::optional<c10::ScalarType> InferExpectedScalarType(const Node* n) {
...
+          // Mimic PyTorch scalar type promotion logic
+          // from https://github.com/pytorch/pytorch/issues/9515
+          if (rank == 0) {
+            auto default_scalar_type =
+                at::typeMetaToScalarType(at::get_default_dtype());
+            switch (scalar_type) {
+              case at::kDouble:
+                typesFromScalars.emplace_back(default_scalar_type);
+                break;
+              case at::kLong:
```
Comment (neginraoof) [inline torch/csrc/jit/passes/onnx/scalar_type_analysis.cpp:123]: Can you specify the behavior here for case Long, maybe in comments?

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 5]
```diff
@@ -3476,7 +3479,7 @@ def forward(self, x):
     def test_scalar_type(self):
         class ArithmeticModel(torch.nn.Module):
             def forward(self, x):
-                return x.size(0) * 2 * x
+                return x.size(0) * 2 * x, 2 - x
```
Comment (neginraoof) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:3482]: Is this test case passing for scripting as well?

**Comment by neginraoof** [comment]
I'm not clear about this item:
ONNX could potentially set inferred shape in output instead of value_infos, checking both to be sure.

Is there any other mechanism to specify shape type other than value_infos? What does it mean to check both?

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 5 above)
Comment (BowenBao) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:3482]: yes it passes locally.

**Comment by BowenBao** [comment]
> 
> 
> I'm not clear about this item:
> ONNX could potentially set inferred shape in output instead of value_infos, checking both to be sure.
> 
> Is there any other mechanism to specify shape type other than value_infos? What does it mean to check both?

Yes, changes in ONNX 1.8 will write inferred shape for outputs of graph to graph outputs. inferred shape for outputs of intermediate node will still be kept in value_infos.

File: torch/csrc/jit/passes/onnx/scalar_type_analysis.cpp (hunk 4 above)
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/scalar_type_analysis.cpp:130]: Added some comments.

File: torch/csrc/jit/passes/onnx/scalar_type_analysis.cpp [hunk 6]
```diff
@@ -108,8 +108,39 @@ static c10::optional<c10::ScalarType> InferExpectedScalarType(const Node* n) {
...
+          auto rank = tensor.dim();
+          auto scalar_type = tensor.scalar_type();
+          // Mimic PyTorch scalar type promotion logic
+          // from https://github.com/pytorch/pytorch/issues/9515
+          // Quoting:
+          //    A Tensor is a considered a "wrapped number" if it is
+          //    auto-wrapped from a C++ or Python number type. Integer types are
+          //    wrapped as 0-dim int64 tensors and floating-point types are
+          //    wrapped as 0-dim double tensors.
+          if (rank == 0) {
```
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/scalar_type_analysis.cpp:121]: yes, the rank is from the tensor data from within constant node though, so yeah this is expected to be present.

File: torch/onnx/utils.py [hunk 7]
```diff
@@ -964,8 +964,21 @@ def _run_symbolic_function(g, n, inputs, env, operator_export_type=OperatorExpor
...
+                    #     block1()
+                    #
+                    # Loop format:
+                    #   prim::Loop(iter, cond, input_1, ..., input_n)
+                    #     block0(iter, input_1, ..., input_n)
+                    #
+                    # For `If` node, there is nothing to copy.
+                    # For `Loop` node, copy metadata for `iter`, `input_1`, ..., `input_n`.
                     for i, b_in in enumerate(b.inputs()):
+                        # if i == 0 and i < len(inputs):
```
Comment (neginraoof) [inline torch/onnx/utils.py:980]: Is this supposed to be commented?

Comment (BowenBao) [reply torch/onnx/utils.py:980]: no, thanks!

**Review by neginraoof — APPROVED** [review]
LGTM. THanks!

**Review by bzinodev — APPROVED** [review]

**Pull request closed without merging**